File: 2001-3000/2302. Count Subarrays With Score Less Than K.java
// Hard
// Slding Window
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/count-subarrays-with-score-less-than-k/

class Solution {
    // count the number of subarrays with ending at right
    public long countSubarrays(int[] nums, long k) {
        long ans = 0, sum = 0;
        int left = 0;

        for (int right = 0; right < nums.length; right++) {
            sum += nums[right];
            while (sum * (right - left + 1) >= k && left <= right) {
                sum -= nums[left];
                left++;
            }

            ans += (right - left + 1);
        }
        return ans;
    }
}

File: 2001-3000/2302. Count Subarrays With Score Less Than K.py
# Time:O(n), Space:O(1)

from typing import List

class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        n = len(nums)
        left = 0
        cur_sum = 0
        ans = 0

        for right in range(n):
            cur_sum += nums[right]

            while cur_sum * (right - left + 1) >= k:
                cur_sum -= nums[left]
                left += 1
                # nothing in window
                if left > right:
                    break
            
            ans += right - left + 1
            
        return ans 
    